crates/l2/contracts/src/l1/based/SequencerRegistry.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.29;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "../interfaces/ISequencerRegistry.sol";
+import "./interfaces/IOnChainProposer.sol";
+
+contract SequencerRegistry is
+    ISequencerRegistry,
+    Initializable,
+    UUPSUpgradeable,
+    OwnableUpgradeable
+{
+    uint256 public constant MIN_COLLATERAL = 1 ether;
+
+    uint256 public constant BATCHES_PER_SEQUENCER = 32;
+
+    address public ON_CHAIN_PROPOSER;
+
+    mapping(address => uint256) public collateral;
+    address[] public sequencers;
+
+    function initialize(
+        address owner,
+        address onChainProposer
+    ) public initializer {
+        require(
+            onChainProposer != address(0),
+            "SequencerRegistry: Invalid onChainProposer"
+        );
+        ON_CHAIN_PROPOSER = onChainProposer;
+
+        _validateOwner(owner);
+        OwnableUpgradeable.__Ownable_init(owner);
+    }
+
+    function register(address sequencer) public payable {
+        _validateRegisterRequest(sequencer, msg.value);
+
+        collateral[sequencer] = msg.value;
+        sequencers.push(sequencer);
+
+        emit SequencerRegistered(sequencer, msg.value);
+    }
+
+    function unregister(address sequencer) public {
+        _validateUnregisterRequest(sequencer);
+
+        uint256 amount = collateral[sequencer];
+        collateral[sequencer] = 0;
+        for (uint256 i = 0; i < sequencers.length; i++) {
+            if (sequencers[i] == sequencer) {
+                sequencers[i] = sequencers[sequencers.length - 1];
```

Comment:
Will this affect round robin?

Reply:
Now, it affects both past, present and future. It is not a problem for this stage (it will be in the next step, the P2P introduction, but we'll already tackling this).